synapse/handlers/oidc_handler.py

```diff
@@ -149,26 +150,29 @@ async def handle_oidc_callback(self, request: SynapseRequest) -> None:
         # otherwise, it is presumably a successful response. see:
         #   https://tools.ietf.org/html/rfc6749#section-4.1.2
 
-        # Fetch the session cookie
+        # Fetch the session cookie. See notes in handle_redirect_request about why
+        # we have two of these.
         session = request.getCookie(SESSION_COOKIE_NAME)  # type: Optional[bytes]
+        if session is None:
+            session = request.getCookie(SESSION_COOKIE_NAME_IOS_HACK)
         if session is None:
             logger.info("Received OIDC callback, with no session cookie")
             self._sso_handler.render_error(
                 request, "missing_session", "No session cookie found"
             )
             return
 
-        # Remove the cookie. There is a good chance that if the callback failed
+        # Remove the cookies. There is a good chance that if the callback failed
         # once, it will fail next time and the code will already be exchanged.
-        # Removing it early avoids spamming the provider with token requests.
-        request.addCookie(
-            SESSION_COOKIE_NAME,
-            b"",
-            path="/_synapse/oidc",
-            expires="Thu, Jan 01 1970 00:00:00 UTC",
-            httpOnly=True,
-            sameSite="lax",
-        )
+        # Removing the cookies early avoids spamming the provider with token requests.
+        for cookie_name, options in [
+            (SESSION_COOKIE_NAME, b"; SameSite=None"),
+            (SESSION_COOKIE_NAME_IOS_HACK, b""),
+        ]:
+            request.cookies.append(
+                b"%s=; Path=/_synapse/client/oidc; Expires=Thu, Jan 01 1970 00:00:00 UTC; "
+                b"HttpOnly; Secure%s" % (cookie_name, options)
+            )
```

Comment:
This is to get around Twisted not supporting setting `SameSite=None`?

Reply:
yes.